File: server/routes/lessonPlans.ts
/**
 * Documents: reading, writing, sharing and deleting them.
 *
 * "Lesson plan" is the storage name for what the rest of the app calls a
 * document — the table predates the editor. Access is either ownership or
 * co-authorship, and that rule is spelled out once (`writableBy`) rather than
 * at each of the four handlers that need it: it had been copy-pasted as a raw
 * `jsonb @>` fragment, and a rule duplicated four times is a rule that only
 * has to be fixed in three places to become a hole.
 */

import { Hono } from "hono";
import { zValidator } from "@hono/zod-validator";
import { and, desc, eq, inArray, or, sql } from "drizzle-orm";

import { getUser } from "../kinde";
import { db } from "../db";
import { idParam } from "../http";
import { requireDb } from "../middleware";
import {
  lessonPlans as lessonPlanTable,
  lessonPlanInputSchema,
} from "../db/schema/lessonPlans";
import { users as userTable } from "../db/schema/users";

/**
 * Documents this user may read and write: the ones they own, plus the ones
 * they are named on as a co-author.
 *
 * `coAuthors` is a jsonb array of user ids, so membership is a containment
 * test rather than a join — there is no join table to keep in step.
 */
function writableBy(userId: string) {
  return or(
    eq(lessonPlanTable.userId, userId),
    sql`${lessonPlanTable.coAuthors} @> ${JSON.stringify([userId])}::jsonb`,
  );
}

/**
 * Look up usernames for a set of user ids, in one query.
 *
 * Returns a lookup rather than a parallel array so a caller with several
 * documents resolves every co-author across all of them at once — the list
 * endpoint used to issue this query per document.
 */
async function usernameLookup(ids: string[]): Promise<(id: string) => string> {
  if (!ids.length) return id => id;

  const rows = await db!
    .select({ id: userTable.id, username: userTable.username })
    .from(userTable)
    .where(inArray(userTable.id, [...new Set(ids)]));

  const byId = new Map(rows.filter(row => row.username).map(row => [row.id, row.username!]));
  // An id with no username (or no user) stands in for itself, so a co-author
  // shows as *something* rather than vanishing from the list.
  return id => byId.get(id) ?? id;
}

const coAuthorsOf = (plan: { coAuthors: string[] | null }) => plan.coAuthors ?? [];

export const lessonPlansRoute = new Hono()
  // Every handler below touches the database.
  .use("*", requireDb)

  // ── Everything you can edit ──
  .get("/", getUser, async (c) => {
    const rows = await db!
      .select({ lessonPlan: lessonPlanTable, authorUsername: userTable.username })
      .from(lessonPlanTable)
      .leftJoin(userTable, eq(lessonPlanTable.userId, userTable.id))
      .where(writableBy(c.var.user.id))
      .orderBy(desc(lessonPlanTable.createdAt))
      .limit(100);

    // One lookup for the whole page rather than one per document.
    const usernameOf = await usernameLookup(rows.flatMap(row => coAuthorsOf(row.lessonPlan)));

    const lessonPlans = rows.map(({ lessonPlan, authorUsername }) => ({
      ...lessonPlan,
      authorUsername,
      coAuthorUsernames: coAuthorsOf(lessonPlan).map(usernameOf),
    }));

    return c.json({ lessonPlans });
  })

  // ── Published documents ──
  // Ahead of "/:id" so the literal segment is not read as an id.
  .get("/public", async (c) => {
    const lessonPlans = await db!
      .select()
      .from(lessonPlanTable)
      .where(eq(lessonPlanTable.isPublic, true))
      .orderBy(desc(lessonPlanTable.createdAt))
      .limit(100);
    return c.json({ lessonPlans });
  })

  .get("/public/:id", async (c) => {
    const id = idParam(c.req.param("id"));
    if (id === null) return c.json({ error: "Invalid lesson plan ID" }, 400);

    const [lessonPlan] = await db!
      .select()
      .from(lessonPlanTable)
      .where(and(eq(lessonPlanTable.id, id), eq(lessonPlanTable.isPublic, true)))
      .limit(1);

    // A private document and a missing one answer alike: whether an id exists
    // is not something an unauthenticated caller should be able to probe.
    if (!lessonPlan) return c.json({ error: "Public lesson plan not found" }, 404);
    return c.json(lessonPlan);
  })

  // ── One document ──
  .get("/:id", getUser, async (c) => {
    const id = idParam(c.req.param("id"));
    if (id === null) return c.json({ error: "Invalid lesson plan ID" }, 400);

    const [row] = await db!
      .select({ lessonPlan: lessonPlanTable, authorUsername: userTable.username })
      .from(lessonPlanTable)
      .leftJoin(userTable, eq(lessonPlanTable.userId, userTable.id))
      .where(and(eq(lessonPlanTable.id, id), writableBy(c.var.user.id)))
      .limit(1);

    if (!row) return c.json({ error: "Lesson plan not found" }, 404);

    const coAuthors = coAuthorsOf(row.lessonPlan);
    const usernameOf = await usernameLookup(coAuthors);

    return c.json({
      ...row.lessonPlan,
      authorUsername: row.authorUsername,
      coAuthorUsernames: coAuthors.map(usernameOf),
    });
  })

  // ── Create ──
  .post("/", getUser, zValidator("json", lessonPlanInputSchema), async (c) => {
    // Validated by zValidator; the owner comes from the session, never the body.
    const [created] = await db!
      .insert(lessonPlanTable)
      .values({ ...c.req.valid("json"), userId: c.var.user.id })
      .returning();
    return c.json(created, 201);
  })

  // ── Replace ──
  .put("/:id", getUser, zValidator("json", lessonPlanInputSchema), async (c) => {
    const id = idParam(c.req.param("id"));
    if (id === null) return c.json({ error: "Invalid lesson plan ID" }, 400);

    // One statement, not two. The access check used to run as its own SELECT
    // with the same WHERE clause the UPDATE then repeated — which is both a
    // wasted round trip and a window in which access could change between
    // them. An UPDATE that matches nothing returns nothing, which is the same
    // answer the SELECT was being asked for.
    const [updated] = await db!
      .update(lessonPlanTable)
      .set({ ...c.req.valid("json"), updatedAt: new Date() })
      .where(and(eq(lessonPlanTable.id, id), writableBy(c.var.user.id)))
      .returning();

    if (!updated) return c.json({ error: "Lesson plan not found" }, 404);
    return c.json(updated);
  })

  // ── Delete ──
  // Owner only: a co-author may edit a document but not destroy it.
  .delete("/:id", getUser, async (c) => {
    const id = idParam(c.req.param("id"));
    if (id === null) return c.json({ error: "Invalid lesson plan ID" }, 400);

    const [deleted] = await db!
      .delete(lessonPlanTable)
      .where(and(eq(lessonPlanTable.id, id), eq(lessonPlanTable.userId, c.var.user.id)))
      .returning({ id: lessonPlanTable.id });

    if (!deleted) return c.json({ error: "Lesson plan not found" }, 404);
    return c.body(null, 204);
  });

File: backend/services/lesson_plans.py
"""Documents: reading, writing, sharing and deleting them.

Access is either ownership or co-authorship, and that rule is spelled out once
— in ``writable_by`` — rather than at each of the four operations that need it.
It had been copy-pasted as a raw ``jsonb @>`` fragment, and a rule duplicated
four times is a rule that only has to be fixed in three places to become a hole.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal, Optional, Sequence

from sqlalchemy import ColumnElement, delete, or_, select, text
from sqlalchemy.ext.asyncio import AsyncSession

from ..core.errors import NotFound
from ..db.models import LessonPlan, User
from ..schemas.lesson_plan import LessonPlanInput, LessonPlanOut
from .users import username_lookup

#: How many documents one listing returns.
LIST_LIMIT = 100


def writable_by(user_id: str) -> ColumnElement[bool]:
    """Documents this user may read and write.

    The ones they own, plus the ones they are named on as a co-author.
    ``co_authors`` is a jsonb array of user ids, so membership is a containment
    test rather than a join — there is no join table to keep in step.
    """
    return or_(
        LessonPlan.user_id == user_id,
        LessonPlan.co_authors.op("@>")(text("CAST(:co_author_id AS jsonb)")).params(
            co_author_id=f'["{user_id}"]'
        ),
    )


def _co_authors(plan: LessonPlan) -> list[str]:
    return list(plan.co_authors or [])


async def _decorate(
    db: AsyncSession, rows: Sequence[tuple[LessonPlan, Optional[str]]]
) -> list[LessonPlanOut]:
    """Attach author and co-author usernames to a page of documents.

    One lookup for the whole page rather than one per document.
    """
    resolve = await username_lookup(db, [uid for plan, _ in rows for uid in _co_authors(plan)])
    out = []
    for plan, author_username in rows:
        model = LessonPlanOut.model_validate(plan)
        model.authorUsername = author_username
        model.coAuthorUsernames = [resolve(uid) for uid in _co_authors(plan)]
        out.append(model)
    return out


async def list_writable(db: AsyncSession, user_id: str) -> list[LessonPlanOut]:
    """Everything this user can edit."""
    rows = (
        (
            await db.execute(
                select(LessonPlan, User.username)
                .outerjoin(User, LessonPlan.user_id == User.id)
                .where(writable_by(user_id))
                .order_by(LessonPlan.created_at.desc())
                .limit(LIST_LIMIT)
            )
        )
        .tuples()
        .all()
    )
    return await _decorate(db, rows)


async def list_public(db: AsyncSession) -> Sequence[LessonPlan]:
    return (
        (
            await db.execute(
                select(LessonPlan)
                .where(LessonPlan.is_public.is_(True))
                .order_by(LessonPlan.created_at.desc())
                .limit(LIST_LIMIT)
            )
        )
        .scalars()
        .all()
    )


#: What a given viewer may do with a document they have the link to.
#:
#: ``owner`` and ``co-author`` both mean "may edit", and are kept apart because
#: only the owner may delete or publish it. ``reader`` is the one that exists
#: for shared links: someone who is on nobody's list but followed a URL to a
#: published document.
Access = Literal["owner", "co-author", "reader"]


async def get_shared(
    db: AsyncSession, plan_id: int, user_id: Optional[str]
) -> tuple[LessonPlanOut, Access]:
    """One document as a *link* reaches it, with what this viewer may do.

    The three paths to a document used to be three requests the client made in
    turn — try it as mine, fall back to trying it as published — which meant
    the client decided what access it had by watching which fetch failed, and
    a private document answered 404 twice before anything was shown. Deciding
    it here is one query and one answer.

    A document that exists but is neither writable nor published is reported as
    missing, exactly as a document that does not exist is: whether an id is
    real is not something a stranger with a URL should be able to probe.
    """
    row = (
        (
            await db.execute(
                select(LessonPlan, User.username)
                .outerjoin(User, LessonPlan.user_id == User.id)
                .where(LessonPlan.id == plan_id)
                .limit(1)
            )
        )
        .tuples()
        .first()
    )
    if row is None:
        raise NotFound("Lesson plan not found")

    plan = row[0]
    if user_id is not None and plan.user_id == user_id:
        access: Access = "owner"
    elif user_id is not None and user_id in _co_authors(plan):
        access = "co-author"
    elif plan.is_public:
        access = "reader"
    else:
        raise NotFound("Lesson plan not found")

    return (await _decorate(db, [row]))[0], access


def _values(payload: LessonPlanInput, *, creating: bool) -> dict[str, Any]:
    """The columns a write should set.

    ``is_public`` and ``co_authors`` are only included when the caller actually
    said something about them. A PUT here writes every key it is handed, so
    including a field unconditionally means an omitted one is written as its
    default — which is how published documents were withdrawn by their own
    autosave, and how the publish toggle (which sends neither field) deleted
    every collaborator on the document it published.
    """
    values: dict[str, Any] = {
        "name": payload.name,
        "main_topic": payload.mainTopic,
        "topics": [topic.model_dump() for topic in payload.topics],
    }
    if payload.coAuthors is not None:
        values["co_authors"] = payload.coAuthors
    elif creating:
        values["co_authors"] = []
    if payload.isPublic is not None:
        values["is_public"] = payload.isPublic
    elif creating:
        # A new document is private until someone says otherwise.
        values["is_public"] = False
    return values


async def create(db: AsyncSession, payload: LessonPlanInput, user_id: str) -> LessonPlan:
    """The owner comes from the session, never from the body."""
    plan = LessonPlan(user_id=user_id, **_values(payload, creating=True))
    db.add(plan)
    await db.flush()
    await db.refresh(plan)
    return plan


async def replace(
    db: AsyncSession, plan_id: int, payload: LessonPlanInput, user_id: str
) -> LessonPlan:
    """Overwrite a document the caller may write, or 404.

    The access check is the same statement as the write. It used to run as its
    own SELECT with an identical WHERE clause, which is both a wasted round
    trip and a window in which access could change between the two.
    """
    plan = (
        await db.execute(
            select(LessonPlan).where(LessonPlan.id == plan_id, writable_by(user_id)).limit(1)
        )
    ).scalar_one_or_none()
    if plan is None:
        raise NotFound("Lesson plan not found")

    for column, value in _values(payload, creating=False).items():
        setattr(plan, column, value)
    plan.updated_at = datetime.utcnow()
    await db.flush()
    await db.refresh(plan)
    return plan


async def remove(db: AsyncSession, plan_id: int, user_id: str) -> None:
    """Delete a document. Owner only — a co-author may edit but not destroy."""
    result = await db.execute(
        delete(LessonPlan).where(LessonPlan.id == plan_id, LessonPlan.user_id == user_id)
    )
    if not result.rowcount:
        raise NotFound("Lesson plan not found")
